### PR TITLE
Remove deprecated code (estimator, safety oracle)

### PR DESCRIPTION
**Pull request by nzpr**
## Overview
This is the first part in series of changes (`finalization` feature branch) aiming to replace current finalizer with the new one that supports block merge. 
in this PR no new code is introduced, only several tests are adjusted that were relying on finalization + dummy values are placed where finalizer output is required. PTAL at TODO comments next to adjustments.

Changes list:
1. Remove storage for last finalized block. This should be reimplemented as part of integration of new finalizer.
2. Remove obsolete finalizer.
3. Remove Estimator and safety oracle. As new version of finalization does not rely on main parent computation, this code is obsolete.
4. Adjust tests. Tests that do not make sense anymore are ignored. Tests that assumes some specific finalizer output are adjusted to not make such assumption.  

### Notes
One tests remaining failing is integration test for finalization. It does not make sense to ignore it and as this PR is based on a feature branch, it can and should be fixed later when finalization code is merged.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala [hunk 1]
```diff
@@ -36,7 +36,7 @@ object CasperLaunch {
     /* Transport */   : TransportLayer: CommUtil: BlockRetriever: EventPublisher
     /* State */       : EnvVars: EngineCell: RPConfAsk: ConnectionsCell: LastApprovedBlock
     /* Rholang */     : RuntimeManager
-    /* Casper */      : Estimator: SafetyOracle: LastFinalizedHeightConstraintChecker: SynchronyConstraintChecker
+    /* Casper */      :  LastFinalizedHeightConstraintChecker: SynchronyConstraintChecker
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/engine/CasperLaunch.scala]: Extra space on multiple _engine files_.

File: node/src/main/scala/coop/rchain/node/api/DeployGrpcServiceV1.scala [hunk 2]
```diff
@@ -143,17 +143,6 @@ object DeployGrpcServiceV1 {
...
-      def showMainChain(request: BlocksQuery): Observable[BlockInfoResponse] =
-        Observable
-          .fromTask(
-            deferCollection(BlockAPI.showMainChain[F](request.depth, apiMaxBlocksLimit)) { r =>
-              import BlockInfoResponse.Message
-              import BlockInfoResponse.Message._
-              BlockInfoResponse(r.fold[Message](Error, BlockInfo))
-            }
-          )
-          .flatMap(Observable.fromIterable)
```
Comment (tgrospic) [inline node/src/main/scala/coop/rchain/node/api/DeployGrpcServiceV1.scala:155]: ~Why is this removed?~ NVM. I see now it's the main parent which doesn't make sense in multi-parent finalizer.

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala [hunk 3]
```diff
@@ -206,21 +163,25 @@ class MultiParentCasperImpl[F[_]
...
-      dag         <- BlockDagStorage[F].getRepresentation
-      r           <- Estimator[F].tips(dag, approvedBlock)
-      (lca, tips) = (r.lca, r.tips)
-
-      /**
-        * Before block merge, `EstimatorHelper.chooseNonConflicting` were used to filter parents, as we could not
-        * have conflicting parents. With introducing block merge, all parents that share the same bonds map
-        * should be parents. Parents that have different bond maps are only one that cannot be merged in any way.
-        */
+      dag <- BlockDagStorage[F].getRepresentation
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:166]: You can use `blockDag` function

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 4]
```diff
@@ -192,7 +193,8 @@ object BlockAPI {
         implicit casper: MultiParentCasper[F]
     ): F[ApiErr[(Seq[DataWithBlockInfo], Int)]] =
       for {
-        mainChain           <- getMainChainFromTip[F](depth)
+        // TODO adjust LFDAN for multiparent or replace with ExploreDeploy
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: What is LFDAN?

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala [hunk 5]
```diff
@@ -192,7 +193,8 @@ object BlockAPI {
...
+        mainChain           <- IndexedSeq.empty[BlockMessage].pure
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Why changing to ~~correct~~ empty sequence is correct?

I think it is better to change implementation of `getMainChainFromTip` to return empty sequence.

File: casper/src/main/scala/coop/rchain/casper/engine/GenesisCeremonyMaster.scala [hunk 6]
```diff
@@ -51,7 +51,7 @@ object GenesisCeremonyMaster {
     /* Transport */   : TransportLayer: CommUtil: BlockRetriever: EventPublisher
     /* State */       : EngineCell: RPConfAsk: ConnectionsCell: LastApprovedBlock
     /* Rholang */     : RuntimeManager
-    /* Casper */      : Estimator: SafetyOracle: LastFinalizedHeightConstraintChecker: SynchronyConstraintChecker
+    /* Casper */      :  LastFinalizedHeightConstraintChecker: SynchronyConstraintChecker
```
Comment (stanislavlyalin) [inline casper/src/main/scala/coop/rchain/casper/engine/GenesisCeremonyMaster.scala]: Extra space also

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala [hunk 7]
```diff
@@ -432,69 +416,6 @@ class MultiParentCasperAddBlockSpec extends FlatSpec with Matchers with Inspecto
     }
   }
```
Comment (stanislavlyalin) [inline casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: I cannot detect why this test was removed. Is it necessary?

File: casper/src/test/scala/coop/rchain/casper/api/ListeningNameAPITest.scala [hunk 8]
```diff
@@ -13,13 +13,15 @@ import coop.rchain.p2p.EffectsTestInstances.LogicalTime
...
+// TODO All tests depend on main parent chain which is not present in new multiparent Casper.
+//  Adjust after implementation of new finalizer or remove if LFDAN is replaced wit Explore deploy.
 class ListeningNameAPITest extends FlatSpec with Matchers with Inside {
 
   import GenesisBuilder._
 
   val genesis = buildGenesis()
 
-  "getListeningNameDataResponse" should "work with unsorted channels" in effectTest {
+  "getListeningNameDataResponse" should "work with unsorted channels" ignore effectTest {
```
Comment (stanislavlyalin) [inline casper/src/test/scala/coop/rchain/casper/api/ListeningNameAPITest.scala]: Why such many tests are ignored? Because broken changes?

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala (hunk 4 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Listen For Data At Name API method.
https://github.com/rchain/rchain/blob/65d20ccf0b97bf3f24df18d50ee9c30c8ce0f6e5/node/src/main/scala/coop/rchain/node/api/DeployGrpcServiceV1.scala#L172

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala (hunk 5 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: It's fixed in https://github.com/rchain/rchain/pull/3661/commits/eb783368b6efc0f7a8f32578d9a837120f418449. `getMainChainFromTip` is completely removed.

File: casper/src/test/scala/coop/rchain/casper/api/ListeningNameAPITest.scala (hunk 8 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/api/ListeningNameAPITest.scala]: Yes, those that depend on estimator and safety oracle (single parent finalizer).

With the fix for LFDAN in this file are not disabled.

File: casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala (hunk 3 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala:166]: You are right but we want to remove this file completely so it's not so important.

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala (hunk 7 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: I don't see why it should be removed, it runs correctly.

Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: I reverted deletion of this test https://github.com/rchain/rchain/pull/3661/commits/238f03221648551c6a6ea05c9fe28b0d34ef92bc.

**Comment by tgrospic** [comment]
I changed base branch to dev so that BlockStore and BlockAPI changes can be combined together.

To enable tests to pass integration test for finalization is temporary ignored https://github.com/rchain/rchain/pull/3661/commits/2f8aa0e119fc0df0776fe9a69a53207eee9bead8.

**Review by tgrospic — APPROVED** [review]
LGTM

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala (hunk 5 above)
Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: I did this because there should not be a main chain at all, because this requires Estimator which is obsolete.
@tgrospic thanks, nice workaround, though we still should not have any notion of main chain, is it correct? So am I right that this change is just to make tests happy for now and being able to merge to dev?

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala (hunk 7 above)
Comment (nzpr) [reply casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: I'm not sure why to keep it, as we do not have Estimator at all, so nothing to test.

**Review by stanislavlyalin — APPROVED** [review]
LGTM

File: casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala (hunk 5 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala]: Yes, no main chain. It's a fix for API to be compatible after HF2. LFDAN should still work the same.

File: casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala (hunk 7 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala]: OK, we can remove it later.

**Comment by tgrospic** [comment]
bors merge
